### PR TITLE
Add visibility section to pre-publish panel

### LINKED ISSUE #43705: Add the collapsible Visibility section
**Description**

Use the sidebar/modal to show some product settings from the Organization tab.
Note that gutenberg also has something similar for posts and pages already, see the [`PostVisibility`](https://github.com/WordPress/gutenberg/blob/0a5bbb4a07bab5b0f2ee0a1276339eda5d1a5a95/packages/editor/src/components/post-visibility/index.js#L20) component, we should try to re-use / mimic as much as we can from this component.

**Epic**

#40371

Depends on #43704

**Design (temporary link)**

[AkVNGImLgSqCObTQ3idVn7-fi-3359_344908](https://href.li/?https://www.figma.com/file/AkVNGImLgSqCObTQ3idVn7/?node-id=3359-344908)

**ACs**

```[tasklist]
### Tasks
- [ ] Create the collapsible Visibility section AkVNGImLgSqCObTQ3idVn7-fi-3359_345477
- [ ] Add `Hide in product catalog` checkbox like in the Organization tab
- [ ] Add `Hide from search results` checkbox like in the Organization tab
- [ ] Add `Enable product reviews` checkbox like in the Organization tab
- [ ] Add `Require a password` checkbox like in the Organization tab
- [ ] When the `Require a password` box is checked, the user can enter the password directly in the panel
- [ ] The section and the block in the Organization tab are fully synced
- [ ] The title `Visibility: Public` changes to `Visibility: Hidden` if `Hide in product catalog` or `Hide from search results` as checked (there’s also a badge in the title bar)
```

### PR DESCRIPTION
**Pull request by octaedro**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds a visibility section to the pre-publish panel.

Screenshot 2024-02-08 at 09 50 16

Closes #43705.

**ACs**

```[tasklist]
### Tasks
- [ ] Create the collapsible Visibility section AkVNGImLgSqCObTQ3idVn7-fi-3359_345477
- [ ] Add `Hide in product catalog` checkbox like in the Organization tab
- [ ] Add `Hide from search results` checkbox like in the Organization tab
- [ ] Add `Enable product reviews` checkbox like in the Organization tab
- [ ] Add `Require a password` checkbox like in the Organization tab
- [ ] When the `Require a password` box is checked, the user can enter the password directly in the panel
- [ ] The section and the block in the Organization tab are fully synced
- [ ] The title `Visibility: Public` changes to `Visibility: Hidden` if `Hide in product catalog` or `Hide from search results` as checked
```

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure the New product editor and the Prepublish sidebar (product-pre-publish-modal) are enabled under `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`.
2. Go to Products > Add New
3. Add a product `Name` and press `Add`.
4. The prepublish panel should be visible.
5. Verify that the visibility section is present in the panel.
6. Check the different options and verify they are synced with the options under the `Organization` tab.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/blocks/generic/checkbox/edit.tsx [hunk 1]
```diff
@@ -39,7 +39,7 @@ export function Edit( {
 	return (
 		<div { ...blockProps }>
 			<Checkbox
-				value={ value || null }
+				value={ value || false }
```
Comment (octaedro) [inline packages/js/product-editor/src/blocks/generic/checkbox/edit.tsx:42]: I changed this because when `value = false`, we used to send `null` to the `Checkbox` component. That was problematic on multiple levels.

File: packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+				visibility: catalogVisibility,
+			} );
+		}
+	}
+
+	return (
+		<PanelBody
+			initialOpen={ false }
+			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+			// @ts-ignore We need to send an Element.
```
Comment (octaedro) [inline packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx:38]: Considering how [the `title` prop is being used](https://github.com/WordPress/gutenberg/blob/62a7316473b74911670fc28fcfe3b927a7fbb808/packages/editor/src/components/post-publish-panel/prepublish.js#L112-L120) on the GB pre-publish panel, I think that this [TS rule](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/panel/types.ts#L93) will change. Otherwise we should modify it in the future.

File: packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+			title={ [
+				__( 'Visibility:', 'woocommerce' ),
+				<span className="editor-post-publish-panel__link" key="label">
+					{ catalogVisibility === 'visible'
+						? __( 'Public', 'woocommerce' )
+						: __( 'Hidden', 'woocommerce' ) }
```
Comment (louwie17) [inline packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx:44]: We should create a separate issue for adding `Password protected` as the visibility status here when a user enters a password. Cause at the moment it would still display `Public` which is confusing.

File: packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+							'woocommerce'
+						) }
+						checked={
+							catalogVisibility === 'catalog' ||
+							catalogVisibility === 'hidden'
+						}
+						onChange={ ( selected ) =>
+							handleVisibilityChange( selected, 'catalog' )
+						}
+					/>
```
Comment (louwie17) [inline packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx:70]: The above two checkboxes and the `handleVisibilityChange` function has a lot of overlap with the catalog visibility block: https://github.com/woocommerce/woocommerce/blob/793e4a821ddda5ca286a857338a02063ba33df0c/packages/js/product-editor/src/blocks/product-fields/catalog-visibility/edit.tsx#L30

Could we instead move the logic from the catalog visibility block to a re-useable component and use that in replacement of the above two checkboxes. That way the business logic around the catalog visibility stays in one place.

File: packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+		'postType',
+		productType,
+		'post_password'
+	);
+
+	const { requirePassword } = useDispatch( productEditorUiStore );
+
+	const isPasswordRequired: boolean = useSelect( ( select ) => {
+		return select( productEditorUiStore ).isPasswordRequired();
+	}, [] );
```
Comment (louwie17) [inline packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx]: I don't understand the need for having `requirePassword` being part of the `productEditorUiStore`, this seems unnecessary when we can just look at the `postPassword`. `isPasswordRequired = Boolean( postPassword )`.

Relying on the data store can cause a discrepancy between the product data and the data store, we should just rely on the product data.

File: packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+							id={ postPasswordId }
+							label={ __( 'Password', 'woocommerce' ) }
+						>
+							<InputControl
+								id={ postPasswordId }
+								value={ postPassword }
+								onChange={ setPostPassword }
+							/>
+						</BaseControl>
+					) }
```
Comment (louwie17) [inline packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx]: The password logic seems identical to the password block: https://github.com/woocommerce/woocommerce/blob/4ae79618c6d250555ce9e54ef5af55d35fb98bed/packages/js/product-editor/src/blocks/product-fields/password/edit.tsx#L50
Similar to the catalog block I think we should create a `ProductPassword` component and re-use it in the block and here.

File: packages/js/product-editor/src/blocks/product-fields/password/edit.tsx [hunk 7]
```diff
@@ -32,26 +34,31 @@ export function Edit( {
...
 	const postPasswordId = useInstanceId(
 		BaseControl,
 		'post_password'
 	) as string;
 
+	const { requirePassword } = useDispatch( productEditorUiStore );
+
+	const isPasswordRequired: boolean = useSelect( ( select ) => {
+		return select( productEditorUiStore ).isPasswordRequired();
+	}, [] );
```
Comment (louwie17) [inline packages/js/product-editor/src/blocks/product-fields/password/edit.tsx]: I mentioned this in a comment further below, but I am not sure on the reasoning behind this change and reliance on the new data store. This currently breaks the form and adds an unnecessary layer, I think we should revert these changes and remove the changes in the productEditorUiStore.

**Review by louwie17 — CHANGES_REQUESTED** [review]
This tested well, although I did notice one issue around the password logic and I left an inline comment around reverting most of that.
Otherwise this looks good, I do think we can consolidate some more of this logic and I left a couple more inline comments around that as well.

File: packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx (hunk 5 above)
Comment (octaedro) [reply packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx]: @louwie17 I added this to address the AC that requires full syncing between the visibility panel and the Organization tab (_The section and the block in the Organization tab are fully synced_).

We need to control whether the `Require a password` checkbox is checked in the pre-publish panel or under the Organization tab, to sync them.

`isPasswordRequired = Boolean( postPassword )` won't help us to sync both checkboxes because it only will be `true` after setting the password, leading us to the situation when the `Require a password` checkbox could be checked in one place, but unchecked in another place.

I'm not opposed to modifying that AC (_The section and the block in the Organization tab are fully synced_), but I would confirm with @j111q before because it could be odd for the user that 3 of 4 checkboxes are synced and that one not.

File: packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx (hunk 4 above)
Comment (octaedro) [reply packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx:70]: Good idea! I did it in the commits https://github.com/woocommerce/woocommerce/pull/44434/commits/47ffe61dadd32ded8e9fc902faa0452035130480 and https://github.com/woocommerce/woocommerce/pull/44434/commits/70e74cbba6740bf8192833e643d777bdc941727a

File: packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx (hunk 6 above)
Comment (octaedro) [reply packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx]: Done in commit the https://github.com/woocommerce/woocommerce/pull/44434/commits/78e507e896325e38cefef3dd3ec5dadf99545d37

File: packages/js/product-editor/src/components/catalog-visibility/catalog-visibility.tsx [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+			/>
+			<CheckboxControl
+				label={ __( 'Hide from search results', 'woocommerce' ) }
+				checked={
+					catalogVisibility === 'catalog' ||
+					catalogVisibility === 'hidden'
+				}
+				onChange={ ( selected ) =>
+					handleVisibilityChange( selected, 'catalog' )
+				}
```
Comment (louwie17) [inline packages/js/product-editor/src/components/catalog-visibility/catalog-visibility.tsx]: How come you changed this to two checkboxes? couldn't you keep this as a singular checkbox the same as the block worked as a singular checkbox?

Comment (louwie17) [reply packages/js/product-editor/src/components/catalog-visibility/catalog-visibility.tsx]: And pass the `visibility` prop in.

File: plugins/woocommerce/src/Internal/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php [hunk 9]
```diff
@@ -509,24 +509,9 @@ private function add_organization_group_blocks() {
...
-				'id'         => 'product-catalog-catalog-visibility',
-				'blockName'  => 'woocommerce/product-catalog-visibility-field',
-				'order'      => 30,
-				'attributes' => array(
-					'label'      => __( 'Hide from search results', 'woocommerce' ),
-					'visibility' => 'catalog',
-				),
+				'id'        => 'product-catalog-search-visibility',
+				'blockName' => 'woocommerce/product-catalog-visibility-fields',
+				'order'     => 20,
```
Comment (louwie17) [inline plugins/woocommerce/src/Internal/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php]: I don't like how we combined this into a single block, I believe we should keep this as before and just have it render a single checkbox. See the comment in regarding the `CatalogVisibility` by removing the additional checkbox.

File: packages/js/product-editor/src/components/require-password/require-password.tsx [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+		'post_password'
+	) as string;
+
+	const [ isPasswordRequired, setPasswordRequired ] = useState(
+		Boolean( postPassword )
+	);
+
+	useEffect( () => {
+		setPasswordRequired( postPassword !== '' );
+	}, [ postPassword ] );
```
Comment (louwie17) [inline packages/js/product-editor/src/components/require-password/require-password.tsx:44]: This logic causes the weird behaviour where if you delete all the text it automatically unchecks and hides the password field ( I mentioned this in Slack ). I think we should avoid that, this would be the trickiest part about this.

I originally thought we could rely on the `isSelected` value that is passed into a block and use that to decipher if a user is using the block to update the password or not, but it appears that `isSelected` is only set to `false` when click on another block and not when I click on the pre-publish modal.

Instead to keep it simple, lets change the logic to this:
```
if ( ! isPasswordRequired && postPassword !== '' ) {
  setPasswordRequired( true );
}
```
So that it only gets checked automatically, but not unchecked automatically.


https://github.com/woocommerce/woocommerce/assets/2240960/2a97f2ac-fa04-4b5f-8bb7-faf2995a7c5b

**Review by louwie17 — CHANGES_REQUESTED** [review]
Thanks for consolidating some of the logic into their own components. I did leave one comment around the catalog visibility as I wasn't suggesting to merge both of the checkboxes together as well, I think we should keep those separate in the same way the block was working.
Also suggested a slight logic change to the password uncheck logic, which may be a good middle ground for now around the syncing.

File: packages/js/product-editor/src/components/catalog-visibility/catalog-visibility.tsx (hunk 8 above)
Comment (octaedro) [reply packages/js/product-editor/src/components/catalog-visibility/catalog-visibility.tsx]: Now I get the idea! I changed that in the commit https://github.com/woocommerce/woocommerce/pull/44434/commits/69d9ee5b51bf223c164c984805db03f519aea955

**Review by louwie17 — APPROVED** [review]
LGTM 🚀 Thanks for the updates Fernando, this is testing well and the code changes look good :shipit:
